src/work/Work.cpp

```diff
@@ -1,437 +1,161 @@
-    mApp.getMetrics().NewMeter({"work", "unit", "run"}, "unit").Mark();
-    setState(WORK_RUNNING);
-    onRun();
+    return !mChildren.empty();
 }
 
-void
-Work::complete(CompleteResult result)
+bool
+Work::anyChildRaiseFailure() const
 {
-    CLOG(DEBUG, "Work") << "completed " << getUniqueName();
-    auto& succ =
-        mApp.getMetrics().NewMeter({"work", "unit", "success"}, "unit");
-    auto& fail =
-        mApp.getMetrics().NewMeter({"work", "unit", "failure"}, "unit");
-
-    switch (result)
-    {
-    case WORK_COMPLETE_OK:
-        setState(onSuccess());
-        break;
-    case WORK_COMPLETE_FAILURE:
-        setState(WORK_FAILURE_RETRY);
-        break;
-    case WORK_COMPLETE_FATAL:
-        setState(WORK_FAILURE_FATAL);
-        break;
-    }
-
-    switch (getState())
+    for (auto& c : mChildren)
     {
-    case WORK_SUCCESS:
-        succ.Mark();
-        CLOG(DEBUG, "Work")
-            << "notifying parent of successful " << getUniqueName();
-        notifyParent();
-        break;
-
-    case WORK_FAILURE_RETRY:
-        fail.Mark();
-        onFailureRetry();
-        scheduleRetry();
-        break;
-
-    case WORK_FAILURE_RAISE:
-    case WORK_FAILURE_FATAL:
-        fail.Mark();
-        onFailureRaise();
-        CLOG(DEBUG, "Work") << "notifying parent of failed " << getUniqueName();
-        notifyParent();
-        break;
-
-    case WORK_PENDING:
-        succ.Mark();
-        advance();
-        break;
-
-    case WORK_RUNNING:
-        succ.Mark();
-        scheduleRun();
-        break;
-
-    default:
-        assert(false);
-        break;
+        if (c->getState() == BasicWork::WORK_FAILURE_RAISE)
+        {
+            return true;
+        }
     }
-}
-
-void
-Work::onReset()
-{
-}
-
-void
-Work::onStart()
-{
-}
-
-void
-Work::onRun()
-{
-    scheduleSuccess();
-}
-
-Work::State
-Work::onSuccess()
-{
-    return WORK_SUCCESS;
-}
-
-void
-Work::onFailureRetry()
-{
-}
-
-void
-Work::onFailureRaise()
-{
-}
-
-Work::State
-Work::getState() const
-{
-    return mState;
+    return false;
 }
 
 bool
-Work::isDone() const
-{
-    return mState == WORK_SUCCESS || mState == WORK_FAILURE_RAISE ||
-           mState == WORK_FAILURE_FATAL;
-}
-
-void
-Work::setState(Work::State st)
+Work::anyChildFatalFailure() const
 {
-    auto maxR = getMaxRetries();
-    if (st == WORK_FAILURE_RETRY && (mRetries >= maxR))
+    for (auto& c : mChildren)
     {
-        CLOG(WARNING, "Work")
-            << "Reached retry limit " << maxR << " for " << getUniqueName();
-        st = WORK_FAILURE_RAISE;
-    }
-
-    if (st != mState)
-    {
-        CLOG(DEBUG, "Work") << "work " << getUniqueName() << " : "
-                            << stateName(mState) << " -> " << stateName(st);
-        mState = st;
+        if (c->getState() == BasicWork::WORK_FAILURE_FATAL)
+        {
+            return true;
+        }
     }
+    return false;
 }
 
-void
-Work::notifyParent()
+std::shared_ptr<Work>
+Work::yieldNextRunningChild()
```

Comment:
I totally agree with you, but for this prototype my goal is to exhibit how the interface would roughly work, without focusing on particular optimizations. 
To your point though, this will probably get addressed in the actual implementation.